### PR TITLE
Dynamic sparsity pattern iterator

### PR DESCRIPTION
**Pull request by bangerth**
This PR adds iterators over the elements of a `DynamicSparsityPattern`. The `SparsityPattern` class already has such iterators, and I need the ones here in order to get rid of the deprecated `*SparsityPattern::row_iterator` types that expose the underlying data structures of the sparsity pattern classes.

### CODE REVIEW

File: include/deal.II/lac/dynamic_sparsity_pattern.h [hunk 1]
```diff
@@ -38,6 +39,203 @@ template <typename number> class SparseMatrix;
...
+   * The typical use for these iterators is to iterate over the elements of a
+   * sparsity pattern (or, since they also serve as the basis for iterating
+   * over the elements of an associated matrix, over the elements of a sparse
+   * matrix), or over the elements of individual rows. There is no
+   * guarantee that the elements of a row are actually traversed in an order
+   * in which column numbers monotonically increase. See the documentation of the
+   * SparsityPattern class for more information.
+   *
+   * @note This class operates directly on the internal data
+   * structures of the SparsityPattern class. As a consequence, some
```
Comment (kronbichler) [inline include/deal.II/lac/dynamic_sparsity_pattern.h]: should read DynamicSparsityPattern here.

**Comment by tamiko** [comment]
I will have a look early next week ;-)

**Comment by kronbichler** [comment]
Regarding the old 'row_begin' and 'row_end' methods: Do you plan to also deprecate them on the dynamic sparsity pattern where they are currently not marked as 'deprecated'?

The reason I am asking is regarding efficiency: We use this method in SparsityPattern::copy_from and a few other places. However, the nature of the new Accessor class makes it somewhat expensive to iterate over the rows only because the new check
`while (index_within_row == sparsity_pattern->lines[current_row].entries.size())` is executed on every increment operator and thus whenever we access the column entries. This is going to show up in some way or another, at least on weaker CPUs without good branch predictors. At least internally to deal.II I do see some value in having a faster access to the column data in dynamic sparsity pattern.

**Comment by kronbichler** [comment]
Apart from that, I think it is a good idea to unify the externally visible interfaces.

**Comment by bangerth** [comment]
@kronbichler : yes, I had imagined I would remove these functions. I think they were a mistake to begin with.  Do you think that this will make things *significantly* more expensive? As far as I recall we only use the existing iterators to copy sparsity patterns but these are operations one does only once, so I think it is relatively cheap. Furthermore, I'm really only comparing one integer against the size of a vector which will typically require the subtraction of two pointers. I imagine that what we do with the result of the iterator would usually cost far more.

Instead of storing the index within the array for each line, I could just store a pointer into it and then compare with `entries.end ()`. Would you prefer this?

**Comment by bangerth** [comment]
@tamiko : Chill out. I'm not purposely holding your patch back. I took care of maybe 50 emails yesterday (a Sunday!) that had piled up over the course of the last week while I was working on aspect things while Timo was here and then for a three day workshop.  I essentially spent the entire rest of the week chasing emails, doing patch review, and other things I do for others. You need to allow me the occasional hour to program something for myself or I will burn out. I know that I'm not always perfect in getting everything done in the most timely manner but it's not for lack of trying. It's because I need a few hours doing things for myself every week so I can recover.

**Comment by tamiko** [comment]
> @tamiko : Chill out.

I thought one smiley ";-)" would be enough to explicitly mark the
comment as humoristic. :-/

Sorry for the miscommunication.

I hope your e-mail workload goes down a bit again.

Best,
Matthias

**Comment by kronbichler** [comment]
> Instead of storing the index within the array for each line, I could just store a pointer into it and then compare with `entries.end ()`. Would you prefer this?

That would certainly make the job easier because it avoids the pointer difference and possibly be easier for the compiler to figure out, so I would think that this is definitely a way forward. When this patch is merged I will benchmark the versions at some point to see how it affects us. It might be fine because many operations done on sparsity patterns are memory-intensive but since my patch in #760 I actually increased the use of DynamicSparsityPattern. (If things turn out bad for reorder_Cuthill_McKee, I can change things back if necessary.)

**Comment by bangerth** [comment]
@kronbichler : I've rewritten the iterators to store a pointer into the current row instead of the index. I've also addressed the typo.

Right now there is nothing for you to time since I'm not using these iterators anywhere outside the testsuite, but that is of course the next step.

**Comment by kronbichler** [comment]
This looks good. I will do some timing manually and provide in a few minutes.

**Comment by kronbichler** [comment]
Here is some timing (clang 3.5.0, release build) with step-22 in 3D, cycle 4 (328k DoFs):

Current implementation based on `DynamicSparsityPattern::row_iterator`:
```
DoFRenumbering::Cuthill_McKee:                     3.87s
SparsityTools::reorder_hierarchical on block(0,0): 31.07s
BlockSparsityPattern::copy_from(dsp):              0.140s
```

When changing the implementation to `DynamicSparsityPattern::iterator` in `sparsity_tools.cc` and `sparsity_pattern.cc`:
```
DoFRenumbering::Cuthill_McKee:                     3.95s
SparsityTools::reorder_hierarchical on block(0,0): 31.99s
BlockSparsityPattern::copy_from(dsp):              0.234s
```

So the increase in computing time in reordering algorithms which are heavy users of the iterators in DynamicSparsityPattern is very small, a few percent only. `copy_from` is affected a bit more but I don't consider it acceptable, considering that even the updated time is more than twenty times less the time for the actual creation of the pattern.

Bottom line: I fully support this plan because it simplifies things a lot and does not affect performance in a measurable way.

(Don't mind the relatively slow performance of my reorder_hierarchical method here: The sparsity pattern contains many rows of length one (constrained dofs) which are not relevant in the grid reordering context and have not been optimized for; on those cases the performance is within a factor of 2 of Cuthill_McKee.)

**Comment by bangerth** [comment]
OK, thanks for checking. I will go ahead with removing the old iterators then.
